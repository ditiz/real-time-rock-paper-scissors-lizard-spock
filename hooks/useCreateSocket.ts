import { useEffect, useState } from "react"
import { Socket } from "socket.io"
import SocketIOClient from "socket.io-client"

const useCreateSocket = () => {
  const [connected, setConnected] = useState<boolean>(false)
  const [nbUser, setNbUser] = useState<number>(0)
  const [socket, setSocket] = useState<Socket>()

  useEffect(() => {
    const url = window.location.host
    const clientSocket = SocketIOClient(url, {
      path: "/api/socket/init",
    })

    setSocket(clientSocket as unknown as Socket)

    // log socket connection
    clientSocket.on("connect", () => {
      setConnected(true)
    })

    // Update the number of user when evetn "new-user" is emited
    clientSocket.on("new-user", (option) => {
      setNbUser(option.connectionCount)
    })

    setTimeout(() => clientSocket.emit("test"), 1000)

    return () => {
      setSocket(undefined)
      clientSocket.close()
    }
  }, [])

  return { connected, nbUser, socket }
}

export default useCreateSocket
