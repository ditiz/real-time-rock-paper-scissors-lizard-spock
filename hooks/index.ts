import useCreateSocket from "./useCreateSocket"
// imports

export {
  useCreateSocket,
  // exports
}
