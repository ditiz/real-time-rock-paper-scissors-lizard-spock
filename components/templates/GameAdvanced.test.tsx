import * as hooks from "@/hooks"
import "@testing-library/jest-dom"
import { render, screen } from "@testing-library/react"
import GameAdvancedTemplate from "./GameAdvanced"

jest.mock("@/hooks", () => ({
  useCreateSocket: jest.fn(),
}))

it("renders GameAdvancedTemplate unchanged", () => {
  jest.spyOn(hooks, "useCreateSocket").mockImplementationOnce(() => ({
    connected: true,
    nbUser: 2,
    socket: {
      on: jest.fn(),
    } as any,
  }))

  const { container } = render(<GameAdvancedTemplate />)
  expect(container).toMatchSnapshot()
})

it("should loading", () => {
  jest.spyOn(hooks, "useCreateSocket").mockImplementationOnce(() => ({
    connected: false,
    nbUser: 0,
    socket: {
      on: jest.fn(),
    } as any,
  }))

  render(<GameAdvancedTemplate />)
  expect(screen.getByText(/Loading/)).toBeInTheDocument()
})

it("should contain title", () => {
  jest.spyOn(hooks, "useCreateSocket").mockImplementationOnce(() => ({
    connected: true,
    nbUser: 0,
    socket: {
      on: jest.fn(),
    } as any,
  }))

  render(<GameAdvancedTemplate />)
  expect(
    screen.getByText("Rock Paper Scissors Lizard Spock")
  ).toBeInTheDocument()
})

it("should have waiting message", () => {
  jest.spyOn(hooks, "useCreateSocket").mockImplementationOnce(() => ({
    connected: true,
    nbUser: 1,
    socket: {
      on: jest.fn(),
    } as any,
  }))

  render(<GameAdvancedTemplate />)
  expect(screen.getByText(/waiting/i)).toBeInTheDocument()
})

it("should have game", () => {
  jest.spyOn(hooks, "useCreateSocket").mockImplementationOnce(() => ({
    connected: true,
    nbUser: 2,
    socket: {
      on: jest.fn(),
    } as any,
  }))

  render(<GameAdvancedTemplate />)
  expect(screen.getAllByRole("img")).toHaveLength(6)
})
