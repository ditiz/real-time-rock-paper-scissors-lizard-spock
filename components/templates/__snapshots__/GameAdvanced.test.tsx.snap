// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders GameAdvancedTemplate unchanged 1`] = `
<div>
  <nav
    class="navbar"
  >
    <div
      class="navbar__logo"
    >
      <img
        alt="logo"
        data-nimg="1"
        decoding="async"
        height="25"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
        style="color: transparent;"
        width="25"
      />
    </div>
    <div
      class="navbar__links"
    >
      <a
        href="/"
      >
        Home
      </a>
      <a
        href="/Rules"
      >
        Rules
      </a>
      <a
        href="/Classic"
      >
        Classic
      </a>
    </div>
  </nav>
  <header
    class="header"
  >
    <h1>
      Rock Paper Scissors Lizard Spock
    </h1>
  </header>
  <main>
    <h2
      class="game__title"
    >
      Choose your move
    </h2>
    <article
      class="game"
    >
      <section>
        <img
          alt="rock icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="icons/rock.svg"
          style="color: transparent;"
          width="50"
        />
      </section>
      <section>
        <img
          alt="paper icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="icons/paper.svg"
          style="color: transparent;"
          width="50"
        />
      </section>
      <section>
        <img
          alt="scissors icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="icons/scissors.svg"
          style="color: transparent;"
          width="50"
        />
      </section>
      <section>
        <img
          alt="lizard icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="icons/lizard.svg"
          style="color: transparent;"
          width="50"
        />
      </section>
      <section>
        <img
          alt="spock icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="icons/spock.svg"
          style="color: transparent;"
          width="50"
        />
      </section>
    </article>
  </main>
</div>
`;
